unit-tests/py/common.py

```diff
@@ -0,0 +1,117 @@
+def require(exp):  
+    global n_assertions, n_failed_assertions, test_failed
+    n_assertions += 1
+    if not exp:
+        n_failed_assertions += 1
+        test_failed = True
+        print("Require failed")
+        for line in traceback.format_stack():
+            print(line.strip())
+
+# Receives the resulted value and the expected value and asserts they are equal
+def require_equal(result, expected):
+    global n_assertions, n_failed_assertions, test_failed
+    n_assertions += 1
+    if not result == expected:
+        n_failed_assertions += 1
+        test_failed = True
+        print("Result was: " + result + "\nBut we expected: " + expected)
+        for line in traceback.format_stack():
+            print(line.strip())
+
+# This function should never be reached, it receives the number of assertion that were 
+# skipped if it was reached
+def require_no_reach(skipped):
+    require(False)
+
+# Receives 2 lists and asserts they are identical
+def require_equal_lists(result, expected):
+    global n_assertions, n_failed_assertions, test_failed
+    n_assertions += 1
+    failed = False
+    if not len(result) == len(expected):
+        n_failed_assertions += 1
+        test_failed = True
+        failed = True
+        print("Require equal lists failed due to lists of different sizes:")
+        print("The resulted list has " + str(len(result)) + " elements, but the expected list has " + str(len(expected)) + " elements")
+    i = 0
+    for res, exp in zip(result, expected):
+        n_assertions += 1
+        if not res == exp:
+            n_failed_assertions += 1
+            test_failed = True
+            failed = True
+            print("Require equal lists failed due to unequal elements:")
+            print("The element of index " + str(i) + " in both lists was not equal")
+        i += 1
+    if failed:
+        print("Result list:", end = ' ')
+        print(result)
+        print("Expected list:", end = ' ')
+        print(expected)
+        for line in traceback.format_stack():
+            print(line.strip())
+
+# Receives an exception and asserts its type and message is as expected
+def require_exception(exception, expected_type, expected_msg = ""):
+    require_equal(type(exception), expected_type)
+    if expected_msg:
+        require_equal(str(exception), expected_msg)
+
+# Functions for formatting test cases
+def start(msg):
+    global n_tests, test_failed
+    n_tests += 1
+    test_failed = False
+    print(msg)
+
+def finish():
+    global test_failed, n_failed_tests
+    if test_failed:
+        n_failed_tests += 1
+        print("Test failed")
+    else:
+        print("Test passed")
+    print("\n")
+
+def print_results_and_exit():
+    global n_assertions, n_tests, n_failed_assertions, n_failed_tests
+    if n_failed_assertions:
+        passed = str(n_assertions - n_failed_assertions)
+        print("test cases: " + str(n_tests) + " | " + str(n_failed_tests) + " failed")
```

Comment:
Note that the format has to agree with the expected format in check_log() in run-unit-tests and with the C++ format using Catch

Reply:
I did this according to the explanation that was in run-unit-tests. I believe it fits the format

Reply:
I know it does, and trust you -- I want you to add a comment here that says that the formats must match, so someone doesn't do something thinking they're helping "legibility" (which they might) but in the process break things.

Reply:
Oh, I understand. Will do